File: app/src/main/java/com/dmrandevu/gallery/media/blur/RegionScanner.kt
package com.dmrandevu.gallery.media.blur

import android.media.MediaCodec
import android.media.MediaCodecInfo
import android.media.MediaExtractor
import android.media.MediaFormat
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.withContext
import java.io.File
import java.nio.ByteBuffer

/**
 * First half of the blur: walks the video and asks each [RegionFinder] what to cover, roughly
 * every [BlurTimeline.SAMPLE_PERIOD_MS], producing the timeline the shader later reads.
 *
 * Finders share the decode. Faces and plates want the same frames, and decoding the video twice
 * to ask two questions about it would double the wait for nothing.
 *
 * The video is decoded **straight through, once**. The obvious implementation — seek to each
 * sample instant and grab that frame — costs an exact seek per sample, and an exact seek has to
 * decode from the preceding keyframe every time; measured on a 47 s clip that was 39 s of the
 * 65 s pass. Decoding sequentially and simply dropping the frames between samples turns that
 * into one linear decode, which the hardware does in a couple of seconds.
 *
 * Frames reach ML Kit as the decoder's own YUV images, so nothing is ever copied into a Bitmap.
 */
class RegionScanner {

    /**
     * Scans [input] with every [finder] and returns one timeline covering all of them.
     * [onProgress] is called with 0..1.
     */
    suspend fun scan(
        input: File,
        finders: List<RegionFinder>,
        onProgress: (Float) -> Unit
    ): BlurTimeline =
        withContext(Dispatchers.Default) {
            val extractor = MediaExtractor()
            try {
                extractor.setDataSource(input.absolutePath)
                val track = (0 until extractor.trackCount).firstOrNull { index ->
                    extractor.getTrackFormat(index).getString(MediaFormat.KEY_MIME)
                        ?.startsWith("video/") == true
                } ?: return@withContext BlurTimeline.empty()

                extractor.selectTrack(track)
                val format = extractor.getTrackFormat(track)
                val durationUs = format.longOr(MediaFormat.KEY_DURATION, 0L)
                if (durationUs <= 0L) return@withContext BlurTimeline.empty()

                try {
                    decodeAndDetect(extractor, format, durationUs, finders, onProgress)
                } finally {
                    finders.forEach { it.close() }
                }
            } finally {
                extractor.release()
            }
        }

    private suspend fun decodeAndDetect(
        extractor: MediaExtractor,
        format: MediaFormat,
        durationUs: Long,
        finders: List<RegionFinder>,
        onProgress: (Float) -> Unit
    ): BlurTimeline {
        // One builder per finder: chaining a face box onto a plate track because they happened to
        // overlap would drag one region's blur onto the other's path.
        val builders = finders.map { BlurTimeline.Builder(durationUs) }
        val codec = MediaCodec.createDecoderByType(format.getString(MediaFormat.KEY_MIME)!!)
        // Rotation is metadata; the decoder ignores it, so ML Kit is told about it instead and
        // hands back boxes already in the upright orientation the shader works in.
        val rotation = format.intOr(MediaFormat.KEY_ROTATION, 0)

        try {
            // No output surface: decoding into buffers is what makes getOutputImage able to hand
            // back CPU-readable YUV. Rendering to an ImageReader instead looks tempting but the
            // decoder is free to answer with an opaque vendor buffer, and reading its planes
            // takes the process down with a native abort.
            format.setInteger(
                MediaFormat.KEY_COLOR_FORMAT,
                MediaCodecInfo.CodecCapabilities.COLOR_FormatYUV420Flexible
            )
            codec.configure(format, null, null, 0)
            codec.start()

            val info = MediaCodec.BufferInfo()
            var inputDone = false
            // Decode once at the finest rate anyone asked for, and give each finder its own
            // frames out of that.
            val periodsUs = finders.map { it.samplePeriodMs * 1_000 }
            val nextDueUs = LongArray(finders.size)
            var nextSampleUs = 0L
            val stepUs = periodsUs.min()

            while (true) {
                currentCoroutineContext().ensureActive()

                if (!inputDone) {
                    val index = codec.dequeueInputBuffer(DEQUEUE_TIMEOUT_US)
                    if (index >= 0) {
                        val buffer: ByteBuffer = codec.getInputBuffer(index)!!
                        val size = extractor.readSampleData(buffer, 0)
                        if (size < 0) {
                            codec.queueInputBuffer(
                                index, 0, 0, 0, MediaCodec.BUFFER_FLAG_END_OF_STREAM
                            )
                            inputDone = true
                        } else {
                            codec.queueInputBuffer(index, 0, size, extractor.sampleTime, 0)
                            extractor.advance()
                        }
                    }
                }

                when (val index = codec.dequeueOutputBuffer(info, DEQUEUE_TIMEOUT_US)) {
                    MediaCodec.INFO_TRY_AGAIN_LATER, MediaCodec.INFO_OUTPUT_FORMAT_CHANGED -> Unit
                    else -> {
                        if (index < 0) continue
                        val wanted = info.size > 0 && info.presentationTimeUs >= nextSampleUs
                        if (wanted) {
                            nextSampleUs = info.presentationTimeUs + stepUs
                            // Only valid until the buffer goes back to the decoder, so the
                            // detection has to finish first.
                            codec.getOutputImage(index)?.let { image ->
                                val frame = ScannedFrame(image, rotation)
                                finders.forEachIndexed { i, finder ->
                                    if (info.presentationTimeUs < nextDueUs[i]) return@forEachIndexed
                                    nextDueUs[i] = info.presentationTimeUs + periodsUs[i]
                                    builders[i].addSample(
                                        info.presentationTimeUs,
                                        finder.regionsIn(frame)
                                    )
                                }
                            }
                            onProgress(
                                (info.presentationTimeUs.toFloat() / durationUs).coerceIn(0f, 1f)
                            )
                        }
                        // Frames between samples cost nothing beyond the decode itself.
                        codec.releaseOutputBuffer(index, false)
                    }
                }

                if (info.flags and MediaCodec.BUFFER_FLAG_END_OF_STREAM != 0) break
            }
        } finally {
            runCatching { codec.stop() }
            codec.release()
        }
        return BlurTimeline.of(builders.map { it.build() })
    }

    private companion object {
        const val DEQUEUE_TIMEOUT_US = 10_000L
    }
}

private fun MediaFormat.longOr(key: String, fallback: Long) =
    if (containsKey(key)) getLong(key) else fallback

private fun MediaFormat.intOr(key: String, fallback: Int) =
    if (containsKey(key)) getInteger(key) else fallback

File: ios/DMRandevuGaleri/Media/Blur/RegionScanner.swift
// AVAssetTrack and friends predate Sendable and are safe to hand to the decode queue, which is
// the only place they are touched.
@preconcurrency import AVFoundation
import CoreImage
import CoreVideo
import Foundation
import ImageIO

/// First half of the blur: walks the video and asks each ``RegionFinder`` what to cover, roughly
/// every ``BlurTimeline/samplePeriodMS``, producing the timeline the renderer later reads.
///
/// Finders share the decode. Faces and plates want the same frames, and decoding the video twice
/// to ask two questions about it would double the wait for nothing.
///
/// The video is read **straight through, once**. The obvious implementation — seek to each sample
/// instant and grab that frame — costs an exact seek per sample, and an exact seek has to decode
/// from the preceding keyframe every time; measured on Android with a 47 s clip that was 39 s of
/// the 65 s pass. Reading sequentially and simply dropping the frames between samples turns that
/// into one linear decode, which the hardware does in a couple of seconds.
final class RegionScanner: Sendable {

    struct NoVideoTrackError: Error {}

    /// Where the work happens. `copyNextSampleBuffer` blocks, and a whole video's worth of that
    /// has no business sitting on one of Swift concurrency's few cooperative threads.
    private let queue = DispatchQueue(label: "com.dmrandevu.gallery.scan", qos: .userInitiated)

    /// Scans `url` with every finder and returns one timeline covering all of them.
    /// `onProgress` is called with 0…1.
    func scan(
        url: URL,
        finders: [RegionFinder],
        onProgress: @escaping (Double) -> Void
    ) async throws -> BlurTimeline {
        let asset = AVURLAsset(url: url)
        let tracks = try await asset.loadTracks(withMediaType: .video)
        guard let track = tracks.first else { throw NoVideoTrackError() }

        let duration = try await asset.load(.duration)
        guard let durationUS = duration.microseconds, durationUS > 0 else { return .empty() }

        let transform = try await track.load(.preferredTransform)
        let orientation = Self.orientation(from: transform)

        return try await withCheckedThrowingContinuation { continuation in
            queue.async {
                do {
                    let timeline = try self.read(
                        asset: asset,
                        track: track,
                        orientation: orientation,
                        durationUS: durationUS,
                        finders: finders,
                        onProgress: onProgress
                    )
                    continuation.resume(returning: timeline)
                } catch {
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    private func read(
        asset: AVURLAsset,
        track: AVAssetTrack,
        orientation: CGImagePropertyOrientation,
        durationUS: Int64,
        finders: [RegionFinder],
        onProgress: (Double) -> Void
    ) throws -> BlurTimeline {
        let reader = try AVAssetReader(asset: asset)
        let output = AVAssetReaderTrackOutput(
            track: track,
            // BGRA rather than the decoder's native YUV: both Vision and Core Image take it
            // directly, and the conversion happens in the same hardware pass as the decode.
            outputSettings: [kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA]
        )
        // Nothing outlives the loop iteration that reads it, so there is no reason to copy.
        output.alwaysCopiesSampleData = false
        reader.add(output)
        guard reader.startReading() else {
            throw reader.error ?? NoVideoTrackError()
        }
        defer { reader.cancelReading() }

        // One builder per finder: chaining a face box onto a plate track because they happened to
        // overlap would drag one region's blur onto the other's path.
        let builders = finders.map { _ in BlurTimeline.Builder(videoDurationUS: durationUS) }

        // Decode once at the finest rate anyone asked for, and give each finder its own frames
        // out of that.
        let periodsUS = finders.map { $0.samplePeriodMS * 1_000 }
        var nextDueUS = [Int64](repeating: 0, count: finders.count)
        let stepUS = periodsUS.min() ?? BlurTimeline.samplePeriodMS * 1_000
        var nextSampleUS: Int64 = 0

        while let sample = output.copyNextSampleBuffer() {
            if Task.isCancelled { throw CancellationError() }

            guard let presentationUS = CMSampleBufferGetPresentationTimeStamp(sample).microseconds,
                  presentationUS >= nextSampleUS,
                  let buffer = CMSampleBufferGetImageBuffer(sample) else { continue }
            nextSampleUS = presentationUS + stepUS

            // Only valid until the reader takes the buffer back, so detection has to finish first.
            let frame = ScannedFrame(pixelBuffer: buffer, orientation: orientation)
            for (index, finder) in finders.enumerated() {
                guard presentationUS >= nextDueUS[index] else { continue }
                nextDueUS[index] = presentationUS + periodsUS[index]
                // Deliberately not swallowed: a finder that cannot look at a frame has not
                // established that the frame is clear, and treating that as "nothing to cover"
                // is how an unprotected video gets out.
                let regions = try finder.regions(in: frame)
                builders[index].addSample(timeUS: presentationUS, regions: regions)
            }
            onProgress(min(max(Double(presentationUS) / Double(durationUS), 0), 1))
        }

        if reader.status == .failed, let error = reader.error { throw error }
        return .of(builders.map { $0.build() })
    }

    /// The container's rotation, as the orientation tag Vision and Core Image both speak.
    static func orientation(from transform: CGAffineTransform) -> CGImagePropertyOrientation {
        switch (transform.a, transform.b, transform.c, transform.d) {
        case (0, 1, -1, 0): return .right   // 90° clockwise
        case (0, -1, 1, 0): return .left    // 90° anticlockwise
        case (-1, 0, 0, -1): return .down   // upside down
        default: return .up
        }
    }
}
